### PR TITLE
Test-drive RCPP_USE_UNWIND_PROTECT on all OSes

### LINKED ISSUE #2295: Consider using R_tryEval() instead of Rcpp_eval()
If I understand correctly, this would offer a fail-safe alternative, but errors will most likely be printed to the console instead of captured in the exception object.

@kevinushey: Rcpp moved away from R_TopLevelExec() a year ago (https://github.com/RcppCore/Rcpp/issues/311). What would be other drawbacks of using R_tryEval() instead of Rcpp_eval() if performance is important (https://github.com/RcppCore/Rcpp/issues/578)? Thanks!

### LINKED ISSUE #3871: Conditionally enable Rcpp fast evaluation
I think @lionel- suggests that this is still a bit unstable so we should start testing with it on, but not enable by default for users.

Part of closing this issue will involve adding another item to the travis build matrix so we test with it turned on.

### PR DESCRIPTION
**Pull request by krlmlr**
Use `Rcpp_fast_eval()`. Test also on Windows.

Closes #3871.

Closes #2295.

### CODE REVIEW

**Comment by krlmlr** [comment]
Tests fail again :-(

**Comment by lionel-** [comment]
Maybe because of recently merged changes with `.data[[` unquoting. Or because of recently merged changes to expressions auto-naming.

**Comment by lionel-** [comment]
Will have a look this afternoon.

**Comment by lionel-** [comment]
Rebased on https://github.com/tidyverse/dplyr/commit/e11ddf532c0bf10d1f05974fdd1028d1e8040726, this should fix it.

**Comment by lionel-** [comment]
Segfault again:

```
Error: segfault from C stack overflow
Execution halted
Segmentation fault (core dumped)
```

I have added the unwind-protect builds to allowed failures in the last commit. @krlmlr have I changed the Travis config correctly?

**Comment by krlmlr** [comment]
Stages break allowed failures. I'll try to find a config that works for both cases.

**Comment by krlmlr** [comment]
Looks like it works. The information that a build is allowed to fail is visible only when a build actually fails, see e.g. https://travis-ci.org/tidyverse/dplyr/builds/436329365.

File: inst/include/dplyr/data/DataMask.h [hunk 1]
```diff
@@ -508,7 +508,7 @@ class DataMask {
     get_context_env()["..group_number"] = indices.group() + 1;
 
     // evaluate the call in the data mask
-    SEXP res = Rcpp_eval(expr, data_mask);
+    SEXP res = Rcpp_fast_eval(expr, data_mask);
```
Comment (romainfrancois) [inline inst/include/dplyr/data/DataMask.h]: Is this the only place where we evaluate R code ? 

Is Language::eval directly using the fast_eval ?

Comment (lionel-) [reply inst/include/dplyr/data/DataMask.h]: Language::eval uses fast_eval.

**Comment by hadley** [comment]
Could someone do a little benchmarking to see the impact this has?

**Comment by romainfrancois** [comment]
Sure. I'll pick this up. I'll revise my `branchmark` :package: to use bench

**Pull request merged**